File: frontend/src/lib/paymentTerms.ts
/** PO payment terms -- mirrors backend/app/core/payment_terms.py. Stored
 * as "Advance", "Prepaid", "On Delivery" or "Others: <details>". */
export const PAYMENT_TERM_CHOICES = ['Advance', 'Prepaid', 'On Delivery', 'Others'] as const
export type PaymentTermChoice = (typeof PAYMENT_TERM_CHOICES)[number]

export interface PaymentTerms {
  choice: PaymentTermChoice | ''
  details: string
}

/** Anything that isn't one of the fixed choices (older free-text terms,
 * a supplier quote's wording) reads as Others with that text. */
export function parsePaymentTerms(value: string | null | undefined): PaymentTerms {
  const text = (value ?? '').trim()
  if (!text) return { choice: '', details: '' }
  const fixed = PAYMENT_TERM_CHOICES.find((c) => c !== 'Others' && c.toLowerCase() === text.toLowerCase())
  if (fixed) return { choice: fixed, details: '' }
  if (text.toLowerCase().startsWith('others')) return { choice: 'Others', details: text.slice(6).replace(/^[\s:-]+/, '') }
  return { choice: 'Others', details: text }
}

export function composePaymentTerms({ choice, details }: PaymentTerms): string {
  return choice === 'Others' ? `Others: ${details.trim()}` : choice
}

export function paymentTermsError({ choice, details }: PaymentTerms): string | null {
  if (!choice) return 'Payment terms are required.'
  if (choice === 'Others' && !details.trim()) return 'Describe the payment terms for Others.'
  return null
}

export const PAYMENT_TERM_HINTS: Record<PaymentTermChoice, string> = {
  Advance: 'Paid before delivery.',
  Prepaid: 'Already paid -- this PO is for record only. Finance records the payment made.',
  'On Delivery': 'Paid when the goods arrive.',
  Others: 'Describe the terms.',
}

File: backend/app/core/payment_terms.py
"""PO payment terms (docs/modules/purchase_orders.md Revision 8): one of
Advance, Prepaid (the PO is for record only -- Finance records the
payment already made), On Delivery, or "Others: <details>". Stored as
that text in `payment_terms`."""

ADVANCE = "Advance"
PREPAID = "Prepaid"
ON_DELIVERY = "On Delivery"
OTHERS = "Others"
PAYMENT_TERM_CHOICES = (ADVANCE, PREPAID, ON_DELIVERY)


def normalise_payment_terms(value: str) -> str:
    """Returns the stored form, or raises ValueError (a 422 on the field)."""
    value = (value or "").strip()
    for choice in PAYMENT_TERM_CHOICES:
        if value.lower() == choice.lower():
            return choice
    if value.lower().startswith(OTHERS.lower()):
        details = value[len(OTHERS):].lstrip(" :-").strip()
        if not details:
            raise ValueError("Describe the payment terms for Others.")
        return f"{OTHERS}: {details}"
    raise ValueError("Choose Advance, Prepaid, On Delivery or Others.")
